StringIndexOutOfBoundsException during analysis

```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 1, end 0, length 1
        at java.lang.String.checkBoundsBeginEnd(String.java:3319) ~[?:?]
        at java.lang.String.substring(String.java:1874) ~[?:?]
        at tv.hd3g.selfautorestdoc.RESTEntryPoint.lambda$ensurePaths$4(RESTEntryPoint.java:100) ~[selfautorestdoc-0.2.3.jar:0.2.3]
```

https://github.com/hdsdi3g/selfautorestdoc/blob/5a497228c9b94cc32bf49803f3cb28c7e9b88ecd/src/main/java/tv/hd3g/selfautorestdoc/RESTEntryPoint.java#L100

--- a/API.md
+++ b/API.md
@@ -1,15 +1,15 @@
 # REST API
-## Summary
 
- - [Delete Demo **DELETE** /serverPath/deleteActionControllerPath](#delete-demo-delete-/serverpath/deleteactioncontrollerpath)
- - [Get Demo **GET** /serverPath/getActionController/{textValueVarName}/path?numValue=0](#get-demo-get-/serverpath/getactioncontroller/{textvaluevarname}/pathnumvalue=0)
- - [Patch Demo **PATCH** /serverPath/patchActionControllerPath](#patch-demo-patch-/serverpath/patchactioncontrollerpath)
- - [Post Demo **POST** /serverPath/postActionControllerPath](#post-demo-post-/serverpath/postactioncontrollerpath)
- - [Put Demo **PUT** /serverPath/putActionControllerPath](#put-demo-put-/serverpath/putactioncontrollerpath)
+## Summary <a name="top"></a>
 
+* [Delete Demo **DELETE** `/serverPath/deleteActionControllerPath`](#goto546779669)
+* [Get Demo **GET** `/serverPath/getActionController/{textValueVarName}/path?numValue=0`](#goto3292818944)
+* [Patch Demo **PATCH** `/serverPath/patchActionControllerPath`](#goto3046218300)
+* [Post Demo **POST** `/serverPath/postActionControllerPath`](#goto734994824)
+* [Put Demo **PUT** `/serverPath/putActionControllerPath`](#goto1734623014)
 
-## Delete Demo
-**DELETE** /serverPath/deleteActionControllerPath
+## <a name="goto546779669"></a> Delete Demo
+**DELETE** `/serverPath/deleteActionControllerPath`
 
 deleteActionController
 
@@ -22,10 +22,10 @@ Response: "application/json" {
 
 _Mandatory rights: rightForDelete | alternateRightForDelete_
 
-[Go to the top](#rest-api) &bull; [DemoRestController :: deleteActionController](/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#73)
+[Go to the top](#top) &bull; [DemoRestController :: deleteActionController](https://github.com/hdsdi3g/selfautorestdoc/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#73)
 
-## Get Demo
-**GET** /serverPath/getActionController/{textValueVarName}/path?numValue=0
+## <a name="goto3292818944"></a> Get Demo
+**GET** `/serverPath/getActionController/{textValueVarName}/path?numValue=0`
 
 A comment for get demo
 
@@ -46,10 +46,10 @@ Response: "application/json" {
 }
 ```
 
-[Go to the top](#rest-api) &bull; [DemoRestController :: getActionController](/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#59)
+[Go to the top](#top) &bull; [DemoRestController :: getActionController](https://github.com/hdsdi3g/selfautorestdoc/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#59)
 
-## Patch Demo
-**PATCH** /serverPath/patchActionControllerPath
+## <a name="goto3046218300"></a> Patch Demo
+**PATCH** `/serverPath/patchActionControllerPath`
 
 patchActionController
 
@@ -62,10 +62,10 @@ Response: "application/json" {
 
 _Mandatory rights: rightForPatch & anotherRightForPatch_
 
-[Go to the top](#rest-api) &bull; [DemoRestController :: patchActionController](/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#79)
+[Go to the top](#top) &bull; [DemoRestController :: patchActionController](https://github.com/hdsdi3g/selfautorestdoc/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#79)
 
-## Post Demo
-**POST** /serverPath/postActionControllerPath
+## <a name="goto734994824"></a> Post Demo
+**POST** `/serverPath/postActionControllerPath`
 
 postActionController
 
@@ -94,10 +94,10 @@ Response: "application/json" {
 }
 ```
 
-[Go to the top](#rest-api) &bull; [DemoRestController :: postActionController](/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#49)
+[Go to the top](#top) &bull; [DemoRestController :: postActionController](https://github.com/hdsdi3g/selfautorestdoc/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#49)
 
-## Put Demo
-**PUT** /serverPath/putActionControllerPath
+## <a name="goto1734623014"></a> Put Demo
+**PUT** `/serverPath/putActionControllerPath`
 
 putActionController
 
@@ -108,4 +108,5 @@ Response: "application/json" {
 }
 ```
 
-[Go to the top](#rest-api) &bull; [DemoRestController :: putActionController](/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#66)
+[Go to the top](#top) &bull; [DemoRestController :: putActionController](https://github.com/hdsdi3g/selfautorestdoc/blob/master/src/main/java/tv/hd3g/selfautorestdoc/demo/DemoRestController.java#66)
+

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
     <groupId>tv.hd3g</groupId>
     <artifactId>selfautorestdoc</artifactId>
     <packaging>jar</packaging>
-    <version>0.2.3</version>
+    <version>0.2.4-SNAPSHOT</version>
 
     <name>SelfAutoRESTDoc</name>
     <url>https://github.com/hdsdi3g/selfautorestdoc</url>

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
     <groupId>tv.hd3g</groupId>
     <artifactId>selfautorestdoc</artifactId>
     <packaging>jar</packaging>
-    <version>0.2.4-SNAPSHOT</version>
+    <version>0.3.0</version>
 
     <name>SelfAutoRESTDoc</name>
     <url>https://github.com/hdsdi3g/selfautorestdoc</url>

--- a/src/main/java/tv/hd3g/selfautorestdoc/RESTEntryPoint.java
+++ b/src/main/java/tv/hd3g/selfautorestdoc/RESTEntryPoint.java
@@ -89,6 +89,9 @@ class RESTEntryPoint {
 		        .filter(v -> v != null && v.isBlank() == false)
 		        .map(String::trim)
 		        .map(v -> {
+			        if (v.equals("/")) {
+				        return "";
+			        }
 			        var start = 0;
 			        if (v.startsWith("/")) {
 				        start = 1;
@@ -98,7 +101,8 @@ class RESTEntryPoint {
 				        ends = v.length() - 1;
 			        }
 			        return v.substring(start, ends);
-		        }).collect(Collectors.joining("/", "/", ""));
+		        })
+		        .collect(Collectors.joining("/", "/", ""));
 	}
 
 	/**

--- a/src/main/java/tv/hd3g/selfautorestdoc/RESTMethod.java
+++ b/src/main/java/tv/hd3g/selfautorestdoc/RESTMethod.java
@@ -30,6 +30,7 @@ import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ValueConstants;
 
 import spoon.reflect.CtModel;
 import spoon.reflect.code.CtComment;
@@ -74,6 +75,9 @@ class RESTMethod {
 			        final var type = p.getType();
 			        final var required = a.required();
 			        final var defaultValue = a.defaultValue();
+			        if (ValueConstants.DEFAULT_NONE.equals(defaultValue)) {
+				        return new URLVariable(name, type, required);
+			        }
 			        return new URLVariable(name, type, required, defaultValue);
 		        })
 		        .collect(Collectors.toUnmodifiableList());

--- a/src/main/java/tv/hd3g/selfautorestdoc/SelfAutoRESTDoc.java
+++ b/src/main/java/tv/hd3g/selfautorestdoc/SelfAutoRESTDoc.java
@@ -16,12 +16,15 @@
  */
 package tv.hd3g.selfautorestdoc;
 
+import static java.nio.charset.StandardCharsets.UTF_8;
+
 import java.io.File;
 import java.io.IOException;
 import java.io.PrintWriter;
 import java.util.List;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
+import java.util.zip.CRC32;
 
 import org.springframework.web.bind.annotation.RestController;
 
@@ -42,13 +45,14 @@ public class SelfAutoRESTDoc {
 		}
 	}
 
-	public void writeToMD(final File mdFile) throws IOException {
+	public void writeToMD(final File mdFile, final String baseProjectURL) throws IOException {
 		final var entryPoints = processAnalysis();
 
-		try (final var pw = new PrintWriter(mdFile)) {
+		try (final var pw = new PrintWriter(mdFile, UTF_8)) {
 			pw.print("# REST API" + NEW_LINE);
+			pw.print(NEW_LINE);
 			writeSummary(entryPoints, pw);
-			entryPoints.forEach(ep -> writeContent(ep, pw));
+			entryPoints.forEach(ep -> writeContent(ep, pw, baseProjectURL));
 			pw.print(NEW_LINE);
 		}
 	}
@@ -68,38 +72,40 @@ public class SelfAutoRESTDoc {
 	}
 
 	private static String linkifyTitle(final String title) {
-		return title.trim().replace(' ', '-').toLowerCase()
-		        .replace("?", "")
-		        .replace("!", "")
-		        .replace(",", "")
-		        .replace(";", "")
-		        .replace(":", "")
-		        .replace(".", "");
+		final var c = new CRC32();
+		c.update(title.getBytes(UTF_8));
+		return "goto" + Math.abs(c.getValue());
 	}
 
 	private void writeSummary(final List<RESTEntryPoint> entryPoints, final PrintWriter pw) {
-		pw.print("## Summary" + NEW_LINE);
-		entryPoints.forEach(ep -> {
-			pw.print(NEW_LINE);
-			pw.print(" - [");
-			pw.print(ep.getRequestNames() + " **" + ep.getVerbs() + "** " + ep.getPaths());
-			pw.print("](#");
-			pw.print(linkifyTitle(ep.getRequestNames() + " " + ep.getVerbs() + " " + ep.getPaths()));
-			pw.print(")");
-		});
+		pw.print("## Summary <a name=\"top\"></a>" + NEW_LINE);
 		pw.print(NEW_LINE);
+
+		entryPoints.forEach(ep -> {
+			pw.print("* [");
+			final var link = ep.getRequestNames() + " **" + ep.getVerbs() + "** `" + ep.getPaths() + "`";
+			pw.print(link.trim());
+			pw.print("](#");
+			pw.print(linkifyTitle(ep.getRequestNames() + ep.getVerbs() + ep.getPaths()));
+			pw.print(")");
+			pw.print(NEW_LINE);
+		});
 	}
 
-	private void writeContent(final RESTEntryPoint ep, final PrintWriter pw) {
-		pw.print(NEW_LINE);
+	private void writeContent(final RESTEntryPoint ep, final PrintWriter pw, final String baseProjectURL) {
 		pw.print(NEW_LINE);
 		pw.print("## ");
 		final var rqname = ep.getRequestNames();
+
+		pw.print("<a name=\"");
+		pw.print(linkifyTitle(rqname + ep.getVerbs() + ep.getPaths()));
+		pw.print("\"></a> ");
+
 		if (rqname.isEmpty() == false) {
 			pw.print(rqname + NEW_LINE);
 		}
 		pw.print("**" + ep.getVerbs() + "** ");
-		pw.print(ep.getPaths() + NEW_LINE);
+		pw.print("`" + ep.getPaths() + "`" + NEW_LINE);
 		pw.print(NEW_LINE);
 
 		ep.getMethodComments().forEach(c -> {
@@ -163,15 +169,17 @@ public class SelfAutoRESTDoc {
 			pw.print(NEW_LINE);
 		}
 
-		pw.print("[Go to the top](#rest-api)");
+		pw.print("[Go to the top](#top)");
 		pw.print(" &bull; [" + ep.getControllerSimpleName());
 		pw.print(" :: ");
 		pw.print(ep.getMethodName());
-		pw.print("](/blob/master/");
+
+		pw.print("](" + baseProjectURL + "/");
 		pw.print(ep.getControllerFullPath());
 		pw.print("#");
 		pw.print(ep.getLineMethodInController());
 		pw.print(")");
+		pw.print(NEW_LINE);
 	}
 
 }

--- a/src/main/java/tv/hd3g/selfautorestdoc/mod/ExportRESTDocAPICmdLine.java
+++ b/src/main/java/tv/hd3g/selfautorestdoc/mod/ExportRESTDocAPICmdLine.java
@@ -35,13 +35,15 @@ public class ExportRESTDocAPICmdLine implements ApplicationRunner {
 	private SelfAutoRestDocEndpointsListener selfAutoRestDocEndpointsListener;
 	@Value("${selfautorestdoc.outputfile:API.md}")
 	private File outputFile;
+	@Value("${selfautorestdoc.baseProjectURL:/blob/master}")
+	private String baseProjectURL;
 
 	@Override
 	public void run(final ApplicationArguments args) throws Exception {
 		if (args.getNonOptionArgs().contains("export-rest-doc-api") == false) {
 			return;
 		}
-		selfAutoRestDocEndpointsListener.getSelfAutoRESTDoc().writeToMD(outputFile);
+		selfAutoRestDocEndpointsListener.getSelfAutoRESTDoc().writeToMD(outputFile, baseProjectURL);
 		if (args.getNonOptionArgs().contains("dont-quit-after-done") == false) {
 			System.exit(0);
 		}

